### ISSUE TITLE
Can't start syslog when built with erlang.mk

### ISSUE THREAD

**Issue by barndon** [ninenines/erlang.mk#329]
I have a project that names [syslog](https://github.com/Vagabond/erlang-syslog) as a dependency.  When I build and run my project, syslog fails to start.  For example

```
$ erl -pa _rel/${MY_PROJECT}/lib/*/ebin
Erlang/OTP 18 [erts-7.0] [source] [64-bit] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V7.0  (abort with ^G)
1> syslog:start().
{error,"could not load driver syslog_drv: \"cannot open shared object file: No such file or directory\""}
```

The root of the problem seems to be that erlang.mk compiled syslog_drv.c as "syslog.so" and syslog expects to load "syslog_drv.so".  I have confirmed that renaming the file solves the problem.

**Comment by essen** [comment]
Thanks, leaving the tab open to take a look tomorrow. :-)

**Comment by essen** [comment]
OK it's a mismatch between the defaults of erlang.mk and rebar. Let me see what I can do.

**Comment by essen** [comment]
OK I have a small idea on how to fix this but I will need to make sure it doesn't break anything else, and I have to fix CI so it compiles properly the NIFs to begin with. So I will get back to you after I do this, hope that's OK.

It's all about having a different default if we detect we are not depending on an erlang.mk project.

**Comment by sata** [comment]
I'm also affected by this. Is there anything I can help with to find a solution?

**Comment by essen** [comment]
Currently going through my backlog of issues for paying customers, so you know what you can do. :-)

Otherwise, perhaps try changing autopatch so that it sets the correct C_SRC_OUTPUT variable for rebar-based projects, if none was set.

**Comment by rlipscombe** [comment]
Same problem here. Did anyone get it working with the autopatch workaround suggested by @essen? If so, could you give me a pointer?

In the meantime, I used the following workaround:

```
app::
        mv $(DEPS_DIR)/syslog/priv/syslog.so $(DEPS_DIR)/syslog/priv/syslog_drv.so
```

**Comment by sata** [comment]
@rlipscombe Problem is syslog doesn't have any rebar.config or rebar.config.script files which results in erlang.mk invoking dep_autopatch_gen. If we adapt that target it would imply all deps who gets patched with dep_autopatch_gen will have _drv.so output files. In other words not a general solution.

It works though. 

```make
define dep_autopatch_gen
	printf "%s\n" \
		"ERLC_OPTS = +debug_info" \
		"C_SRC_OUTPUT = $(DEPS_DIR)/$(1)/priv/$(1)_drv.so" \
		"include ../../erlang.mk" > $(DEPS_DIR)/$(1)/Makefile
endef
```

But a better solution in my opinion is to change dep_autopatch2 target from

```make
define dep_autopatch2
	$(call erlang,$(call dep_autopatch_appsrc.erl,$(1))); \
	if [ -f $(DEPS_DIR)/$(1)/rebar.config -o -f $(DEPS_DIR)/$(1)/rebar.config.script ]; then \
		$(call dep_autopatch_fetch_rebar); \
		$(call dep_autopatch_rebar,$(1)); \
	else \
		$(call dep_autopatch_gen,$(1)); \
	fi
endef
```
to

```make
define dep_autopatch2
	$(call erlang,$(call dep_autopatch_appsrc.erl,$(1))); \
	if [ -f $(DEPS_DIR)/$(1)/rebar -o -f $(DEPS_DIR)/$(1)/rebar.config -o -f $(DEPS_DIR)/$(1)/rebar.config.script ]; then \
		$(call dep_autopatch_fetch_rebar); \
		$(call dep_autopatch_rebar,$(1)); \
	else \
		$(call dep_autopatch_gen,$(1)); \
	fi
endef
```

in other words not only look for rebar.config and rebar.config.script but rebar as well. It works too for me. 

@essen what do you think about changing the if to look for rebar as well?

**Comment by essen** [comment]
Ah is that what the problem was?

I don't have much problem with that solution I think. I'm guessing all projects that include rebar are rebar projects.

I also need to change something to make erlang.mk detection higher priority than rebar. Apparently Cowboy is detected as a rebar project.

I also need to look into just making a target to announce this is erlang.mk rather than trying to detect it.

Just food for thoughts. Will take care of testing all this as soon as I setup the few VMs people have contributed recently.

**Comment by essen** [comment]
Tomorrow I will do a full test run before/after the patch to make sure nothing breaks, and pushing the patch if everything is fine.

**Comment by essen** [comment]
Fixed in current master. Thanks everyone!
